lib/repack.c

```diff
@@ -1289,3 +1294,52 @@ repack_index_swap(PG_FUNCTION_ARGS)
 	SPI_finish();
 	PG_RETURN_VOID();
 }
+
+/**
+ * @fn      Datum get_table_and_inheritors(PG_FUNCTION_ARGS)
+ * @brief   Return array containing Oids of parent table and its children.
+ *
+ * get_table_and_inheritors(table)
+ *
+ * @param	table	parent table.
+ * @retval	regclass[]
+ */
+Datum
+repack_get_table_and_inheritors(PG_FUNCTION_ARGS)
+{
+	Oid			parent = PG_GETARG_OID(0);
+	List	   *relations;
+	Datum	   *relations_array;
+	int			relations_array_size;
+	ArrayType  *result;
+	ListCell   *lc;
+	int			i;
+
+	LockRelationOid(parent, AccessShareLock);
+
+	/* Check that parent table exists */
+	if (!SearchSysCacheExists1(RELOID, ObjectIdGetDatum(parent)))
```

Comment:
> Is it possible that the relation is dropped after we got AccessShareLock, or we get AccessShareLock on a dropped relation?

There's a possibility that we get a lock on a dropped relation, since this function will be called on a string literal cast to `regclass` type (`'table_name'::regclass`), here:

```plpgsql
relid = ANY(repack.get_table_and_inheritors($%d::regclass))
```
and also here:
```plpgsql
WHERE c.oid = ANY (repack.get_table_and_inheritors($1::regclass))
```

As I've already mentioned, `regclassin()` (the implementation of `::regclass` cast) doesn't take any locks:

```C
/* We might not even have permissions on this relation; don't lock it. */
result = RangeVarGetRelid(makeRangeVarFromNameList(names), NoLock, false);
```

IMHO It's better to exit early if we find out that relation doesn't exist anymore.